packages/wrangler/src/publish.ts

```diff
@@ -164,14 +157,47 @@ export default async function publish(props: Props): Promise<void> {
     let migrations;
     if (config.migrations.length > 0) {
       // get current migration tag
-      const scripts = await fetchResult<
-        { id: string; migration_tag: string }[]
-      >(`/accounts/${accountId}/workers/scripts`);
-      const script = scripts.find(({ id }) => id === scriptName);
+      let script: { id: string; migration_tag: string } | undefined;
+      if (!props.legacyEnv) {
+        try {
+          if (props.env) {
+            const scriptData = await fetchResult<{
+              script: { id: string; migration_tag: string };
+            }>(
+              `/accounts/${accountId}/workers/services/${scriptName}/environments/${props.env}`
+            );
+            script = scriptData.script;
+          } else {
+            const scriptData = await fetchResult<{
+              default_environment: {
+                script: {
+                  id: string;
+                  migration_tag: string;
+                };
+              };
+            }>(`/accounts/${accountId}/workers/services/${scriptName}`);
+            script = scriptData.default_environment.script;
+          }
+        } catch (err) {
+          if (
+            [10090, 10092].includes((err as { code: number }).code) === false
+          ) {
+            throw err;
+          }
+          // else it's a 404, no script found, and we can proceed
+        }
+      } else {
+        const scripts = await fetchResult<
+          { id: string; migration_tag: string }[]
+        >(`/accounts/${accountId}/workers/scripts`);
+        script = scripts.find(({ id }) => id === scriptName);
+      }
+
       if (script?.migration_tag) {
         // was already published once
+        const scriptMigrationTag = script.migration_tag;
         const foundIndex = config.migrations.findIndex(
-          (migration) => migration.tag === script.migration_tag
+          (migration) => migration.tag === scriptMigrationTag
```

Comment:
I had to do this because typescript was being weird and saying script could be undefined here (but not saying so for any further occurrence of `script.migration_tag` 🤷‍♂️